File: frontend/types/journal.ts
// Journal module types

export const JOURNAL_CATEGORIES = ['accomplished', 'grateful', 'proud', 'annoyed', 'learned'] as const
export type JournalCategory = typeof JOURNAL_CATEGORIES[number]

export const CATEGORY_META: Record<JournalCategory, { label: string; icon: string; color: string }> = {
  accomplished: { label: 'Accomplished', icon: 'i-heroicons-trophy', color: 'text-electric-green' },
  grateful: { label: 'Grateful For', icon: 'i-heroicons-heart', color: 'text-cyber-blue' },
  proud: { label: 'Proud Of', icon: 'i-heroicons-star', color: 'text-warning-orange' },
  annoyed: { label: 'Annoyed By', icon: 'i-heroicons-bolt', color: 'text-danger-red' },
  learned: { label: 'Learned', icon: 'i-heroicons-light-bulb', color: 'text-purple-400' },
}

export interface JournalItem {
  id: number
  category: JournalCategory
  position: number
  content: string
}

export interface JournalItemCreate {
  category: JournalCategory
  position: number
  content: string
}

export interface JournalEntry {
  id: number
  date: string
  mood_score: number | null
  notes: string | null
  items: JournalItem[]
  created_at: string
  updated_at: string | null
}

export interface JournalEntryCreate {
  date: string
  mood_score?: number | null
  notes?: string | null
  items: JournalItemCreate[]
}

export interface JournalEntryUpdate {
  mood_score?: number | null
  notes?: string | null
  items?: JournalItemCreate[]
}

export interface JournalStats {
  total_entries: number
  entries_this_week: number
  entries_this_month: number
  current_streak: number
  longest_streak: number
  avg_mood: number | null
}

export interface CategoryReportData {
  count: number
  items: string[]
}

export interface ReportData {
  categories: Record<JournalCategory, CategoryReportData>
  avg_mood: number | null
  entry_count: number
  current_streak: number
  longest_streak: number
}

export interface JournalReport {
  id: number
  report_type: 'monthly' | 'yearly'
  period: string
  data: ReportData
  entry_count: number
  created_at: string
  updated_at: string | null
}

export interface ReportGenerateRequest {
  report_type: 'monthly' | 'yearly'
  period: string
}

// --- Meditation ---

export interface MeditationSession {
  id: number
  date: string
  started_at: string
  duration_seconds: number | null
  notes?: string
  completed: boolean
  created_at: string
}

export interface MeditationSessionCreate {
  date: string
  notes?: string
}

export interface MeditationSessionUpdate {
  duration_seconds?: number
  notes?: string
  completed?: boolean
}

export interface MeditationStats {
  total_sessions: number
  total_minutes: number
  avg_duration_minutes: number
  longest_session_minutes: number
  current_streak_days: number
}

File: backend/app/schemas/journal.py
"""Journal module schemas for request/response validation."""

from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List, Dict, Any


JOURNAL_CATEGORIES = ["accomplished", "grateful", "proud", "annoyed", "learned"]


# --- Journal Item ---

class JournalItemCreate(BaseModel):
    """Schema for creating a journal item."""
    category: str = Field(..., pattern=r"^(accomplished|grateful|proud|annoyed|learned)$")
    position: int = Field(..., ge=0)
    content: str = Field(..., min_length=1)


class JournalItemResponse(JournalItemCreate):
    """Schema for journal item response."""
    id: int

    model_config = ConfigDict(from_attributes=True)


# --- Journal Entry ---

class JournalEntryCreate(BaseModel):
    """Schema for creating a journal entry."""
    date: str = Field(..., pattern=r"^\d{4}-\d{2}-\d{2}$")
    mood_score: Optional[int] = Field(None, ge=1, le=10)
    notes: Optional[str] = None
    items: List[JournalItemCreate] = []


class JournalEntryUpdate(BaseModel):
    """Schema for updating a journal entry."""
    mood_score: Optional[int] = Field(None, ge=1, le=10)
    notes: Optional[str] = None
    items: Optional[List[JournalItemCreate]] = None


class JournalEntryResponse(BaseModel):
    """Schema for journal entry response."""
    id: int
    date: str
    mood_score: Optional[int] = None
    notes: Optional[str] = None
    items: List[JournalItemResponse] = []
    created_at: str
    updated_at: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


# --- Journal Stats ---

class JournalStatsResponse(BaseModel):
    """Dashboard statistics for journal module."""
    total_entries: int
    entries_this_week: int
    entries_this_month: int
    current_streak: int
    longest_streak: int
    avg_mood: Optional[float] = None


# --- Journal Report ---

class ReportGenerateRequest(BaseModel):
    """Request to generate/refresh a report."""
    report_type: str = Field(..., pattern=r"^(monthly|yearly)$")
    period: str = Field(..., pattern=r"^\d{4}(-\d{2})?$")  # YYYY-MM or YYYY


class CategoryReportData(BaseModel):
    """Per-category data in a report."""
    count: int
    items: List[str]


class ReportData(BaseModel):
    """Structure of report JSON data."""
    categories: Dict[str, CategoryReportData]
    avg_mood: Optional[float] = None
    entry_count: int
    current_streak: int
    longest_streak: int


class JournalReportResponse(BaseModel):
    """Schema for journal report response."""
    id: int
    report_type: str
    period: str
    data: ReportData
    entry_count: int
    created_at: str
    updated_at: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)
